### ISSUE TITLE
Type checking and IntelliSense with JSDoc typed js dependencies

### ISSUE THREAD

**Issue by ahocevar** [Microsoft/TypeScript#29824]
## Search Terms

JSDoc types dpendenies

## Suggestion

Possible alternative to #7546: Maybe type checking from JSDoc annotated dependencies can be achieved more easily in a different way. I have created a [gist](https://gist.github.com/ahocevar/9a7253cb4712e8bf38d75d8ac898e36c) with a demo project that successfully uses a dependency ([ol@6.0.0-beta.1](https://www.npmjs.com/package/ol/v/6.0.0-beta.1) with [fully JSDoc typed sources](openlayers/openlayers#9178). This works, but requires a [quite involved configuration](https://gist.github.com/ahocevar/9a7253cb4712e8bf38d75d8ac898e36c#file-tsconfig-json) and [breaks easily](https://github.com/openlayers/openlayers/pull/9178#issuecomment-461577897) in [more complex projects](https://github.com/openlayers/openlayers/pull/9178#issuecomment-461588832).

Overall, it looks like a much lower hanging fruit than generating `.d.ts` files from JSDoc annotated JavaScript code`. As the above example shows, it works already, but it should work out of the box and more reliably.

## Use Cases

Type checking with pure JavaScript dependencies that are either JSDoc typed directly or published with their JSDoc typed sources, instead of requiring `.d.ts` files, which [cannot be generated from `.js` sources](#7546).

## Examples

Two possible scenarios:

 1. A package with fully JSDoc typed JavaScript files. Unrealistic, because class inheritance is only fully recognized by TypeScript when the ES2015 `class` and `extends` keywords are used, which packages don't usually do, because it limits interoperability.
 1. A package that contains its fully JSDoc typed JavaScript sources. Realistic and feasible, especially when source maps can point to these sources directly. An example for this would be [ol@6.0.0-beta.1](https://www.npmjs.com/package/ol/v/6.0.0-beta.1).

So let's assume the latter case is the more common one. If TypeScript would look for a `tsconfig.json` in dependencies (e.g. `node_modules/ol/tsconfig.json`), that file could have a config option to point to the annotated sources. Maybe something like
```js
{
  "compilerOptions": {
    "jsSources": [
      "src/**/*.js"
    ]
  }
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by daKmoR** [comment]
that sounds like a nice solution... currently this needs to be done "manually" e.g. if I want to get the JSDoc types of `<your-package-name>` then I need to set it up like so:
```js
  "include": [
    "**/*.js",
    "node_modules/<your-package-name>/**/*.js"
  ],
  "exclude": [
    "node_modules/!(<your-package-name>)"
  ]
```

Some more details you can find a the blog bost [Type-Safe Web Components with JSDoc](https://dev.to/dakmor/type-safe-web-components-with-jsdoc-4icf#let-your-users-consume-your-types)

is there anything we can do to help?

**Comment by rbiggs** [comment]
Yup, I hate having to put node modules with JSDoc comments in the `includes` definition. It just feels like a hack. I also consider it brittle. If you move something, this will break.

**Comment by rbiggs** [comment]
I noticed a bug when using a module that consists of several files that import JSDoc types from other files. When such a project has an `index.js` file as the point of export, TypeScript fails to parse the JSDoc types imported from other files. There is a solution but it is a hack: importing one function from the file with the types used in other files. This forces TypeScript to recognize the imported types in the source files. https://github.com/microsoft/TypeScript/issues/33136

**Comment by RangerMauve** [comment]
This is very painful and makes the claim that "typescript supports JSDoc" false in practice.

**Comment by trusktr** [comment]
> This has to work without `.d.ts` files, because these [cannot be generated from `.js` sources](#7546), neither with `tsc` nor reliably with third party applications.

This actually can be done! But it also requires a _very very non-obvious_ configuration. We can achieve it with the `typesVersions` field in `package.json` (not even in `tsconfig.json`!):


```json
  "typesVersions": {
    "*": {
      "src/*": [
        "types/*"
      ]
    }
  },
```


Here is a project with this configuration:

https://github.com/potahtml/pota/blob/6d5d04b2b19e993837b9fe78ac2e866cf1a1015a/package.json#L14-L20

See also the `tsconfig.json` there.

Now, when we run `tsc`, it will read the JS files (with JSDoc annotations) and output `.d.ts` files, allowing the project to be published!

Here's what the published types look like, including declaration source maps:

https://unpkg.com/browse/pota@0.14.133/types/

It would be really great to have this working out of the box though, without any of the potential bad side-effects that might come with the mentioned configuration (if any?), and without the obscure configuration to begin with.
